synapse/lib/cell.py

```diff
@@ -4031,3 +4047,219 @@ async def _runReloadFunc(self, name, func):
+        if valid is False:
+            return False, {'mesg': f'API Key shadow mismatch: {iden}',
+                           'user': user, 'name': udef.get('name')}
+
+        return True, {'kdef': kdef, 'udef': udef}
+
+    async def modUserApiKey(self, iden, key, valu):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist, cannot modify it: {iden}', iden=iden)
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+
+        vals = {
+            'modified': s_common.now()
+        }
+        if key == 'name':
+            vals['name'] = valu
+        else:
+            raise s_exc.BadArg(mesg=f'Cannot set {key} on user API Keys.', name=key)
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
+
+        logger.info(f'Updated HTTP API Key {iden} for {user.name}, set {key}={valu}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return kdef
+
+    async def regenerateUserApiKey(self, iden, duration=None):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist, cannot regenerated it: {iden}', iden=iden)
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+
+        _, token, shadow = await s_passwd.generateApiKey(iden=iden)
+
+        now = s_common.now()
+        vals = {
+            'shadow': shadow,
+            'modified': now
+        }
+        kdef['shadow'] = shadow
+
+        # Duration is either None or use we use the provided destination which needs to meet schema.
+        if duration is None:
+            vals['expref'] = None
+            vals['duration'] = None
+        else:
+            vals['expref'] = now
+            vals['duration'] = duration
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
+
+        logger.info(f'Regenerated HTTP API Key {iden} for {user.name}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return token, kdef
+
+    @s_nexus.Pusher.onPush('user:apikey:update')
+    async def _setUserApiKey(self, user, iden, vals):
+        lkey = user.encode('utf-8') + b'cell:apikey' + iden.encode('utf-8')
+        buf = self.slab.get(lkey, db=self.usermetadb)
+        if buf is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist: {iden}')
```

Comment:
this should also never happen - it would be a race between the top side of one  handler starting and the bottom of the del handler or a user was removed.